src/Cli/Microsoft.TemplateEngine.Cli/TemplatePackageCoordinator.cs

```diff
@@ -556,5 +808,367 @@ private async Task<bool> IsTemplateShortNameAsync(string sourceIdentifier, Cance
+                ISettings settings = global::NuGet.Configuration.Settings.LoadDefaultSettings(currentDirectory);
+                PackageSourceProvider packageSourceProvider = new PackageSourceProvider(settings);
+                defaultSources = packageSourceProvider.LoadPackageSources().Where(source => source.IsEnabled);
+                if (includeNuGetFeed)
+                {
+                    var nuGetFeed = new PackageSource(NugetOrgFeed);
+                    defaultSources = defaultSources.Append(nuGetFeed);
+                }
+            }
+            catch (Exception ex)
+            {
+                throw new Exception($"Failed to load NuGet sources configured for the folder {currentDirectory}", ex);
+            }
+
+            if (additionalSources == null || !additionalSources.Any())
+            {
+                if (!defaultSources.Any())
+                {
+                    throw new Exception("No NuGet sources are defined or enabled");
+                }
+                return defaultSources;
+            }
+
+            List<PackageSource> customSources = new List<PackageSource>();
+            foreach (string source in additionalSources)
+            {
+                if (string.IsNullOrWhiteSpace(source))
+                {
+                    continue;
+                }
+                if (defaultSources.Any(s => s.Source.Equals(source, StringComparison.OrdinalIgnoreCase)))
+                {
+                    Reporter.Output.WriteLine($"Custom source {source} is already loaded from default configuration.");
+                    continue;
+                }
+                PackageSource packageSource = new PackageSource(source);
+                if (packageSource.TrySourceAsUri == null)
+                {
+                    Reporter.Output.WriteLine(string.Format("Failed to load resource {0} message", source));
+                    continue;
+                }
+                customSources.Add(packageSource);
+            }
+
+            IEnumerable<PackageSource> retrievedSources = customSources.Concat(defaultSources);
+            if (!retrievedSources.Any())
+            {
+                throw new Exception("No NuGet sources are defined or enabled");
+            }
+            return retrievedSources;
+        }
+
+        private async Task<NugetPackageMetadata?> GetPackageMetadataFromMultipleFeedsAsync(
+            IEnumerable<PackageSource> sources,
+            NugetApiManager apiManager,
+            string packageIdentifier,
+            string? packageVersion = null,
+            CancellationToken cancellationToken = default)
+        {
+            NugetPackageMetadata? highestVersion = null;
+
+            foreach (PackageSource source in sources)
+            {
+                NugetPackageMetadata? package = await apiManager.GetPackageMetadataAsync(packageIdentifier, packageVersion, source, cancellationToken).ConfigureAwait(false);
```

Comment:
potentially this can be done in parallel too, and then will be faster. See the implementation in dotnet/templating:
https://github.com/dotnet/templating/blob/dec5d8684547b8b3347def807d4545de9d309ea0/src/Microsoft.TemplateEngine.Edge/Installers/NuGet/NugetApiPackageManager.cs#L248-L310